cmake/install_helpers.cmake

```
@@ -44,66 +44,63 @@ endfunction()
 function(ginkgo_install_library name)
     ginkgo_add_install_rpath("${name}" "${ARGN}")
 
-    if (WIN32 OR CYGWIN)
-        # dll is considered as runtime
-        install(TARGETS "${name}"
-            EXPORT Ginkgo
-            LIBRARY DESTINATION "${CMAKE_INSTALL_FULL_LIBDIR}"
-            ARCHIVE DESTINATION "${CMAKE_INSTALL_FULL_LIBDIR}"
-            RUNTIME DESTINATION "${CMAKE_INSTALL_FULL_BINDIR}"
-            )
-    else ()
-        # install .so and .a files
-        install(TARGETS "${name}"
-            EXPORT Ginkgo
-            LIBRARY DESTINATION "${CMAKE_INSTALL_FULL_LIBDIR}"
-            ARCHIVE DESTINATION "${CMAKE_INSTALL_FULL_LIBDIR}"
+    install(TARGETS "${name}"
+        EXPORT Ginkgo
+        LIBRARY
+        DESTINATION "${CMAKE_INSTALL_FULL_LIBDIR}"
+        COMPONENT Runtime
+        RUNTIME
+        DESTINATION "${CMAKE_INSTALL_FULL_BINDIR}"
+        COMPONENT Runtime
+        ARCHIVE
+        DESTINATION "${CMAKE_INSTALL_FULL_LIBDIR}"
+        COMPONENT Development
```

I think it is better to have a Ginkgo_ prefix to avoid nameclashes when installed with FetchContent or as submodules.
```suggestion
        COMPONENT Ginkgo_Development
```

---

Good point on subdir builds, we need to keep that in mind. I would prefer doing this by making the installation itself configurable, like GTest does it, because Development and Runtime seem to be common names, and other libraries might want to install Ginkgo itself with the same component

---

I added a `GINKGO_INSTALL` option which, when `OFF`, disables the corresponding `install` functions and `test_install`, does that make sense?

---

I don't think this fixes the issue. There might still be name conflicts with other libraries, since `GINKGO_INSTALL` is on by default. So users would need to know that they have to set it to `OFF` to prevent these name conflicts. I don't know how we could warn users about that.
But the `GINKGO_INSTALL` option can stay, I would only suggest adding the prefixes.

---

I think the issue is much smaller than it seems. A subdir build using Ginkgo would be broken if Ginkgo wasn't being installed, so `GINKGO_INSTALL` being on by default is a sensible value IMO. GTest differs in the respect that we only use it for testing, not for production. Additionally, component installs are an advanced usage of CMake, by default all components get installed directly. If people use the same naming scheme (Runtime and Development), `cmake --install . --component Runtime` is safe to run and installs the application plus Ginkgo, if that is not intended, we can expect advanced users to set a flag manually.

---

I just mean, you already linked a presentation, which criticizes using generic names. As @pratikvn mentioned, if a user includes ginkgo through `FetchContent` and they directly, or another library they are using, has the same generic names there is a conflict. But we can solve that very easily as mentioned in the video.
And again, I'm not arguing against `GINKGO_INSTALL`, only against the generic names.

---

Ah right thanks, I was going by memory on the component setup. I'll use a prefix

---

In that case we probably don't need `GINKGO_INSTALL`, unless we consider users using Ginkgo as a purely build-time dependency in a subtree build a realistic use case?

---

Yes, but I can't see how users would know that. I think it's better to use the non-generic names and not hope that the user reads our documentation and figures out that they have to disable `GINKGO_INSTALL`.

---

Not sure I follow, can you elaborate? What I mean is that if people do a normal (not component-wise) install of their application, they need to install Ginkgo with it, with separate component names, the packageability is still guaranteed. The GTest use case of it being purely a build and test dependency doesn't seem important.

---

That sounds like we agree to use the non-generic names, right? So maybe my last comment was just badly phrased and you can ignore it.